geopandas/tools/overlay.py

```diff
@@ -225,7 +39,10 @@ def _overlay_intersection(df1, df2):
         left.reset_index(drop=True, inplace=True)
         right = df2.geometry.take(pairs["__idx2"].values)
         right.reset_index(drop=True, inplace=True)
-        intersections = left.intersection(right).buffer(0)
+        intersections = left.intersection(right)
+        intersections = intersections.apply(
+            lambda g: g.buffer(0) if g.type in ["Polygon", "MultiPolygon"] else g
+        )
```

Comment:
We might want to write this without an apply (to benefit from vectorized operations in the future, but so can also be done later)